core/design/src/main/kotlin/edu/stanford/spezi/core/design/views/views/views/button/SuspendButton.kt

```diff
@@ -0,0 +1,93 @@
+package edu.stanford.spezi.core.design.views.views.views.button
+
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import edu.stanford.spezi.core.design.component.Button
+import edu.stanford.spezi.core.design.component.StringResource
+import edu.stanford.spezi.core.design.theme.SpeziTheme
+import edu.stanford.spezi.core.design.theme.ThemePreviews
+import edu.stanford.spezi.core.design.views.views.compositionLocal.LocalProcessingDebounceDuration
+import edu.stanford.spezi.core.design.views.views.model.ViewState
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+private enum class SuspendButtonState {
+    IDLE, DISABLED, DISABLED_AND_PROCESSING
+}
+
+@Composable
+fun SuspendButton(
+    title: StringResource,
+    state: MutableState<ViewState> = remember { mutableStateOf(ViewState.Idle) },
+    action: suspend () -> Unit,
+) {
+    SuspendButton(state, action) {
+        Text(title.text())
+    }
+}
+
+@Composable
+fun SuspendButton(
+    state: MutableState<ViewState> = remember { mutableStateOf(ViewState.Idle) },
+    action: suspend () -> Unit,
+    label: @Composable () -> Unit,
+) {
+    val buttonState = remember { mutableStateOf(SuspendButtonState.IDLE) }
+    val coroutineScope = rememberCoroutineScope()
+    val debounceScope = rememberCoroutineScope()
+    val processingDebounceDuration = LocalProcessingDebounceDuration.current
+    val externallyProcessing = buttonState.value == SuspendButtonState.IDLE && state.value == ViewState.Processing
+
+    Button(
+        enabled = buttonState.value == SuspendButtonState.IDLE && !externallyProcessing,
+        onClick = {
+            if (state.value == ViewState.Processing) return@Button
+            buttonState.value = SuspendButtonState.DISABLED
+
+            val debounceJob = debounceScope.launch {
+                delay(processingDebounceDuration)
+
+                if (isActive) {
+                    buttonState.value = SuspendButtonState.DISABLED_AND_PROCESSING
+                }
+            }
+
+            state.value = ViewState.Processing
+            coroutineScope.launch {
+                runCatching {
+                    action()
+                    debounceJob.cancel()
```

Comment:
`rememberCoroutineScope()` returns the same scope within a composable lifecycle. we are cancelling the other coroutineScope this way, and we would never reach `buttonState.value = SuspendButtonState.IDLE` below, hence we would be having the bug of the button being stuck in disabled state and view state in processing.

---

Compose scope shouldn't be used to execute such suspendable actions but rather for suspendable ui actions. The reason is that the coroutine scope delivered via `rememberCoroutineScope()` will get cancelled in case the view would leave the recomposition, e.g when scrolling away and not in view port anymore, app background foreground, navigating to a different screen and so on. We need a longer living scope, e.g. a viewModelScope. 

I know that we are trying to align to iOS API here, but have a look at the approach that I have followed to address this via `PendingActions` as properties in ui states in view model

Reply:
I refactored this to use a boolean state instead of canceling the job. We may still want to open an issue though to think about how we may want to change this as a whole, since the action would get cancelled, if the composable is moved out of view - similar to how it would cancel on iOS actually - so this would require a general workup rather than just a quick Android fix